src/sources/kafka.rs

```diff
@@ -120,121 +131,92 @@ impl SourceConfig for KafkaSourceConfig {
-
-                            // Add source type
-                            log.insert(log_schema().source_type_key(), Bytes::from("kafka"));
-
-                            let msg_key = msg
-                                .key()
-                                .map(|key| Value::from(String::from_utf8_lossy(key).to_string()))
-                                .unwrap_or(Value::Null);
-                            log.insert(&key_field, msg_key);
-
-                            log.insert(&topic_key, Value::from(msg.topic().to_string()));
-
-                            log.insert(&partition_key, Value::from(msg.partition()));
-
-                            log.insert(&offset_key, Value::from(msg.offset()));
-
-                            let mut headers_map = BTreeMap::new();
-                            if let Some(headers) = msg.headers() {
-                                // Using index-based for loop because rdkafka's `Headers` trait
-                                // does not provide Iterator-based API
-                                for i in 0..headers.count() {
-                                    if let Some(header) = headers.get(i) {
-                                        headers_map.insert(
-                                            header.0.to_string(),
-                                            Bytes::from(header.1.to_owned()).into(),
-                                        );
-                                    }
-                                }
-                            }
-                            log.insert(headers_key, Value::from(headers_map));
-
-                            consumer.store_offset(&msg).map_err(|error| {
-                                emit!(KafkaOffsetUpdateFailed { error });
-                            })?;
-
-                            Ok(event)
                         }
                     }
                 }
-            })
-            // Try `forward` after removing old futures.
-            // Error: implementation of `futures_core::stream::Stream` is not general enough
-            // .forward(
-            //     out.sink_compat()
-            //         .sink_map_err(|error| error!(message = "Error sending to sink.", %error)),
-            // )
-            .for_each(|item| {
-                let mut out = out.clone();
-                async move {
-                    if let Ok(item) = item {
-                        if let Err(error) = out.send(item).await {
-                            error!(message = "Error sending to sink.", %error);
-                        }
-                    }
+                log.insert(&headers_key, Value::from(headers_map));
+
+                if let Err(error) = consumer.store_offset(&msg) {
+                    emit!(KafkaOffsetUpdateFailed { error });
+                    continue;
                 }
-            })
-            .await;
-        Ok(())
-    }))
+
+                let _ = out
```

Comment:
I'm wondering if we should actually try to forward first _before_ committing the offset.

Reply:
For my use case, only committing the offset after successful delivery would be awesome. I run vector in a containerized environment without disk buffering, and as such I currently run the risk of committing offsets on kafka without delivery. I imagine this risk is fairly small though.

Reply:
Agreed 👍 I'll make that change as a separate PR to go out with 0.14.0 to keep the scope of this fix small. I'm curious to hear if there is a reason it was this way though.

Reply:
This is something that'll be fully resolved with the end-to-end acking work. Moving the offset commit here would make it _slightly_ better than it currently is, but would still give no guarantees about messages having made it through the rest of the pipeline successfully. I'm fine to make that change, as long as we're clear it's not going to make a big difference.